peekingduck/pipeline/nodes/model/mask_rcnnv1/mask_rcnn_files/rpn.py

```diff
@@ -0,0 +1,346 @@
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""Region Proposal Network for Faster R-CNN
+
+Modifications include:
+- Removed training related code
+- Removed training related arguments from RegionProposalNetwork class
+    - fg_iou_thresh
+    - bg_iou_thresh
+    - batch_size_per_image
+    - positive_fraction
+- Removed tracing related code
+"""
+
+from typing import Dict, List, Tuple
+import torch
+from torch import nn, Tensor
+from torch.nn import functional as F
+from peekingduck.pipeline.nodes.model.mask_rcnnv1.mask_rcnn_files import (
+    boxes as box_ops,
+    detection_utils as det_utils,
+)
+from peekingduck.pipeline.nodes.model.mask_rcnnv1.mask_rcnn_files.image_list import (
+    ImageList,
+)
+
+
+def permute_and_flatten(layer, N, C, H, W):
+    # type: (Tensor, int, int, int, int) -> Tensor
+    # pylint: disable=invalid-name
+    """Permutes a feature output to be the same format as the labels"""
+    layer = layer.view(N, -1, C, H, W)
+    layer = layer.permute(0, 3, 4, 1, 2)
+    layer = layer.reshape(N, -1, C)
+    return layer
+
+
+def concat_box_prediction_layers(box_cls, box_regression):
+    # type: (List[Tensor], List[Tensor]) -> Tuple[Tensor, Tensor]
+    # pylint: disable=invalid-name
+    """Permutes each feature level output to be the same format as labels and concatenates them on
+    the first dimension"""
+    box_cls_flattened = []
+    box_regression_flattened = []
+    # for each feature level, permute the outputs to make them be in the
+    # same format as the labels. Note that the labels are computed for
+    # all feature levels concatenated, so we keep the same representation
+    # for the objectness and the box_regression
+    for box_cls_per_level, box_regression_per_level in zip(box_cls, box_regression):
+        N, AxC, H, W = box_cls_per_level.shape
+        Ax4 = box_regression_per_level.shape[1]
+        A = Ax4 // 4
+        C = AxC // A
+        box_cls_per_level = permute_and_flatten(box_cls_per_level, N, C, H, W)
+        box_cls_flattened.append(box_cls_per_level)
+
+        box_regression_per_level = permute_and_flatten(
+            box_regression_per_level, N, 4, H, W
+        )
+        box_regression_flattened.append(box_regression_per_level)
+    # concatenate on the first dimension (representing the feature levels), to
+    # take into account the way the labels were generated (with all feature maps
+    # being concatenated as well)
+    box_cls_out = torch.cat(box_cls_flattened, dim=1).flatten(0, -2)
+    box_regression_out = torch.cat(box_regression_flattened, dim=1).reshape(-1, 4)
+    return box_cls_out, box_regression_out
+
+
+class RPNHead(nn.Module):
+    """
+    Adds a simple RPN Head with classification and regression heads
+
+    Args:
+        in_channels (int): number of channels of the input feature
+        num_anchors (int): number of anchors to be predicted
+    """
+
+    # pylint: disable=invalid-name
```

Comment:
refactor to meaningful name. `t` probably just stands for `tensor`.

Reply:
amended to `transformed_feature`